src/ic-certified-map/src/rbtree.rs

```diff
@@ -365,420 +335,392 @@ impl<K: 'static + AsRef<[u8]>, V: AsHashTree + 'static> RbTree<K, V> {
+            if is_red(&h.right.as_ref().unwrap().left) {
+                h.right = Some(rotate_right(h.right.take().unwrap()));
                 h = rotate_left(h);
-                flip_colors(h);
+                flip_colors(&mut h);
             }
             h
         }
 
-        unsafe fn move_red_right<K: 'static + AsRef<[u8]>, V: AsHashTree + 'static>(
-            mut h: *mut Node<K, V>,
-        ) -> *mut Node<K, V> {
-            flip_colors(h);
-            if is_red((*(*h).left).left) {
+        fn move_red_right<K: 'static + AsRef<[u8]>, V: AsHashTree + 'static>(
+            mut h: Box<Node<K, V>>,
+        ) -> Box<Node<K, V>> {
+            flip_colors(&mut h);
+            if is_red(&h.left.as_ref().unwrap().left) {
                 h = rotate_right(h);
-                flip_colors(h);
+                flip_colors(&mut h);
             }
             h
         }
 
         #[inline]
-        unsafe fn min<K: 'static + AsRef<[u8]>, V: AsHashTree + 'static>(
-            mut h: *mut Node<K, V>,
-        ) -> *mut Node<K, V> {
-            while !(*h).left.is_null() {
-                h = (*h).left;
+        fn min<K: 'static + AsRef<[u8]>, V: AsHashTree + 'static>(
+            mut h: &mut Box<Node<K, V>>,
+        ) -> &mut Box<Node<K, V>> {
+            while h.left.is_some() {
+                h = h.left.as_mut().unwrap();
             }
             h
         }
 
-        unsafe fn delete_min<K: 'static + AsRef<[u8]>, V: AsHashTree + 'static>(
-            mut h: *mut Node<K, V>,
-        ) -> *mut Node<K, V> {
-            if (*h).left.is_null() {
-                debug_assert!((*h).right.is_null());
-                Node::delete(h);
+        fn delete_min<K: 'static + AsRef<[u8]>, V: AsHashTree + 'static>(
+            mut h: Box<Node<K, V>>,
+        ) -> NodeRef<K, V> {
+            if h.left.is_none() {
+                debug_assert!(h.right.is_none());
+                drop(h);
                 return Node::null();
             }
-            if !is_red((*h).left) && !is_red((*(*h).left).left) {
+            if !is_red(&h.left) && !is_red(&h.left.as_ref().unwrap().left) {
                 h = move_red_left(h);
             }
-            (*h).left = delete_min((*h).left);
-            (*h).subtree_hash = Node::subtree_hash(h);
-            balance(h)
+            h.left = delete_min(h.left.unwrap());
+            h.update_subtree_hash();
+            Some(balance(h))
         }
 
-        unsafe fn go<K: 'static + AsRef<[u8]>, V: AsHashTree + 'static>(
-            mut h: *mut Node<K, V>,
+        fn go<K: 'static + AsRef<[u8]>, V: AsHashTree + 'static>(
+            mut h: Box<Node<K, V>>,
             key: &[u8],
-        ) -> *mut Node<K, V> {
-            if key < (*h).key.as_ref() {
-                if !is_red((*h).left) && !is_red((*(*h).left).left) {
+        ) -> NodeRef<K, V> {
+            if key < h.key.as_ref() {
+                if !is_red(&h.left) && !is_red(&h.left.as_ref().unwrap().left) {
```

Comment:
I noticed that in this method (but also in other places such as `balance`), `unwrap`s are used that assume that some invariants hold. 

E.g. this line assumes that `h.left` is_some. 

Would it make sense to make these invariants more explicit by e.g. using asserts or `expect` with a description of why the invariant holds? 

I'm aware that this also has been the case before and so this may not be something to address in this PR.

Reply:
I believe the invariant here is that we know that the key is in the tree, so left must be present. Added a debug_assert with a description.

The invariants required in balance are probably related to left-leaning property of the tree. I can make another pass and add more asserts in a separate PR, this will require some research.

Reply:
Great, thanks!